### PR TITLE
Persist run options & set at creation (not claim)

### LINKED ISSUE #2079: Janitor marks runs lost before their time
The [Janitor](https://github.com/OpenFn/lightning/blob/main/lib/lightning/janitor.ex) uses this query to determine which runs are `lost` and marks them as so:

```ex
  @doc """
  Return all runs that have been claimed by a worker before the earliest
  acceptable start time (determined by the longest acceptable run time) but are
  still incomplete. This indicates that we may have lost contact with the worker
  that was responsible for executing the run.
  """
  @spec lost(DateTime.t()) :: Ecto.Queryable.t()
  def lost(%DateTime{} = now) do
    max_run_duration_seconds =
      Application.get_env(:lightning, :max_run_duration_seconds)

    grace_period = Lightning.Config.grace_period()

    oldest_valid_claim =
      now
      |> DateTime.add(-max_run_duration_seconds, :second)
      |> DateTime.add(-grace_period, :second)

    final_states = Run.final_states()

    from(att in Run,
      where: is_nil(att.finished_at),
      where: att.state not in ^final_states,
      where: att.claimed_at < ^oldest_valid_claim
    )
  end
```

Now that we have dynamic max_run_duration being passed to the worker and we allow some runs to run longer than others, we need to refactor this query to account for the variable time limits. 

Possible solutions:
1. Is the allowed run duration for a run stored on the `run` itself as metadata? if so, we could re-write this query to check the run "against itself" --> i.e., `where runs.metadata['max_duration'] + runs.started_at > time.now()`
2. ..?

### LINKED ISSUE #2082: Add `options` column to runs table
See #2079 - we should _persist_ the run options (are logs allowed? should we store dataclips? what's the max duration?) on the runs table.

- this not only enables the janitor's work (it should look at fixed data on the runs table, not dynamically calculate the allowable duration for each run based on sets of similar projects.)
- this also ensures that the options passed to a run or a work order are fixed at the time the run/workorder was CREATED. this is way more stable and predictable.

### PR DESCRIPTION
**Pull request by taylordowns2000**
To fix #2079 and close #2082, this changes the way we determine run options. It stores them in the DB at the time the run was created, rather than calculating them on the fly when a worker claims a run. This makes the application more predictable. Fewer surprises!

## Validation Steps

1. run migrations
2. start server
3. create a run
4. verify that the run options in the db are correct (see timeout and save_dataclips)
5. stop server, set default run timeout to something SUPER SHORT, restart WITHOUT a running worker
6. create a run
7. verify that options are set up correctly in DB.
8. stop server, set default run timeout to something big again, restart WITH a running worker.
9. notice that the worker respects the timeout from when the run was created, not from when the run was claimed.
10. repeat the same process as above, but modify the project settings persist-dataclips option. you'll see that the option that was selected when the run is _created_ is peristed and respected, not the option that was selected when the run was claimed.

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** (`:owner`, `:admin`, `:editor`, `:viewer`) have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [ ] Product has **QA'd** this feature

---

~Required context (pre-reading) to review:~

~- https://github.com/OpenFn/lightning/issues/2079~
~- https://github.com/OpenFn/lightning/issues/2082~

~Note that if~ ✨ we like this idea ✨ , ~next we still need to:~

1. ~remove the "get options" bit that happens during claim~
12. ~ensure that these options are delivered properly via the channel to the worker~
13. ~check to see which options are set by the project admin (e.g., enable dataclip storage) and which options are set by the instance administrator (e.g., max run duration)~

### CODE REVIEW

**Comment by taylordowns2000** [comment]
@stuartc , please take a look here! context in the PR itself

**Comment by elias-ba** [comment]
@taylordowns2000 this looks like a valid way to implement this. I am not sure I can't validate it's relevancy but it seems to me as a better way of dynamically calculating the run options. Maybe wait for @stuartc to confirm that point.

**Review by elias-ba — APPROVED** [review]
@taylordowns2000 this implementation looks good to me (assuming that tests and other bits will come later)

**Review by stuartc — COMMENTED** [review]
This looks good, thanks for the detailed PR.

My only comment here is that I think we should make the run options _more_ than a type. Right now it's defined in `RunWithOptions`.

I think is should become an embedded schema, so using Ecto.Schema.

And then in the Run model, it becomes `embeds_one :options, Run.Option`

This will require some light refactoring on `get_run_options` on `UsageLimiter` on both Lightning and Thunderbolt. 

I believe this change will give the run options more prominence and make 'just dumping stuff in' less of a temptation.

**Comment by stuartc** [comment]
> @taylordowns2000 this implementation looks good to me (assuming that tests and other bits will come later)

Oh and yes, @taylordowns2000 - tests please!

**Review by elias-ba — CHANGES_REQUESTED** [review]
@taylordowns2000 this is a huge set of great changes. Good job man 🎉. I left a few change requests and questions. Most of the changes are typos and the questions are probably my misunderstanding. 👏🏽

File: lib/lightning/extensions/usage_limiter.ex [hunk 1]
```diff
@@ -11,6 +11,10 @@ defmodule Lightning.Extensions.UsageLimiter do
   def limit_action(_action, _context), do: :ok
 
   @impl true
-  def get_run_options(_context),
-    do: [run_timeout_ms: Lightning.Config.default_max_run_duration() * 1000]
+  def get_run_options(context) do
+    [
+      save_dataclips: Lightning.Projects.save_dataclips?(context.project_id),
+      run_timeout_ms: Lightning.Config.default_max_run_duration() * 1000
```
[the remark this answers is not in the document]
Comment (stuartc) [reply lib/lightning/extensions/usage_limiter.ex:17]: Great question, and yes you're correct - the default behaviour is to use the global max run.

File: lib/lightning/config.ex [hunk 2]
```diff
@@ -132,8 +131,8 @@ defmodule Lightning.Config do
   end
 
   @doc """
-  The grace period is 20% of the max run duration and may be used to wait for
-  an additional amount of time after a run was meant to be finished.
+  The grace period is configurable and is be used to wait for an additional
```
[the remark this answers is not in the document]
Comment (stuartc) [reply lib/lightning/config.ex:134]: :+1:

**Review by stuartc — CHANGES_REQUESTED** [review]
@taylordowns2000 🙌 this is great!

I made bunch of small changes, namely:

- Using `Lightning.current_time()` for the janitor so we can test against specific times, like in the future etc.
- Changed the Janitor to use a `Repo.stream` approach so we don't try and select all with 1000s of runs (edge case for sure but possible).
- Removed primary key generation for the new embedded schema, we don't need it and Ecto adds one by default.

One biggish thing before we merge is, what happens to existing runs in flight as this is deployed? Is there a fallback, or perhaps we need a migration to default the `options` column?

And while I can totally see that the work orders put options onto new runs for new and retries - I can't see any tests in work_orders_test?

**Comment by taylordowns2000** [comment]
nice. think we're good now @stuartc

**Review by stuartc — APPROVED** [review]
